shards/sched.go

```diff
@@ -0,0 +1,230 @@
+// has acquired on the semaphore. Every process is either fast or slow. A
+// process starts as fast, but is downgraded to slow after a period of
+// time. Downgrading relies on a process co-operatively deciding to downgrade.
+//
+// We intentionally keep the algorithm simple, but have a general interface to
+// allow improvements as we learn more.
+type scheduler struct {
+	// Exclusive process holds a write lock, search processes hold read locks.
+	mu             sync.RWMutex
+	semInteractive *semaphore.Weighted
+	semBatch       *semaphore.Weighted
+
+	// interactiveDuration is how long we run a search query at interactive
+	// priority before downgrading it to a batch/slow query.
+	interactiveDuration time.Duration
+}
+
+func newScheduler(capacity int64) *scheduler {
+	// Burst upto 1/4 of CPUs for batch. This means we now can use upto 1.25 the
+	// amount of CPU.
+	batchCap := capacity / 4
+	if batchCap == 0 {
+		batchCap = 1
+	}
+
+	return &scheduler{
+		semInteractive: semaphore.NewWeighted(capacity),
+		semBatch:       semaphore.NewWeighted(batchCap),
+
+		interactiveDuration: 5 * time.Second,
+	}
+}
+
+// Acquire blocks until a normal process is created (ie for a search
+// request). See process documentation. It will only return an error if the
+// context expires.
+func (s *scheduler) Acquire(ctx context.Context) (*process, error) {
+	s.mu.RLock()
+
+	// Start in interactive. yieldFunc will switch us to batch. sem can be nil
+	// if we fail while switching to batch. nil value prevents us releasing
+	// twice.
+	sem := s.semInteractive
+
+	if err := sem.Acquire(ctx, 1); err != nil {
+		s.mu.RUnlock()
+		return nil, err
+	}
+
+	return &process{
+		releaseFunc: func() {
+			if sem != nil {
+				sem.Release(1)
+				sem = nil
+			}
+			s.mu.RUnlock()
+		},
+		yieldTimer: newDeadlineTimer(time.Now().Add(s.interactiveDuration)),
+		yieldFunc: func(ctx context.Context) error {
+			if sem != nil {
+				sem.Release(1)
+				sem = nil
+			}
+
+			// Try to acquire batch. Only set sem if we succeed so we know we can
+			// clean it up. If this fails we assume the process will stop running
+			// (ctx has expired).
+			semNext := s.semBatch
+			if err := semNext.Acquire(ctx, 1); err != nil {
+				return err
+			}
+
+			sem = semNext
+			return nil
+		},
+	}, nil
+}
+
+// Exclusive blocks until an exclusive process is created. An exclusive
+// process is the only running process. See process documentation.
+func (s *scheduler) Exclusive() *process {
+	// Exclusive process holds a write lock on mu, which ensures we have no
+	// processes running (search semaphores are empty).
+	//
+	// exclusive processes will never yield, so we leave yieldTimer and
+	// yieldFunc nil.
+	s.mu.Lock()
+	return &process{
+		releaseFunc: func() {
+			s.mu.Unlock()
+		},
+	}
+}
+
+// process represents a running search query or an exclusive process. When the
+// process is done a call to Release is required.
+type process struct {
```

Comment:
Should we give a process a unique ID so that we can distinguish them in the logs?

Reply:
not a bad idea. Would like it to somehow tie back to a search at the request entry level. Just an ID for this purpose may not be useful, but lets see how our logs look and then it may be justified :)